flink-runtime/src/main/java/org/apache/flink/runtime/metrics/dump/MetricDumpSerialization.java

```diff
@@ -124,50 +124,86 @@ public MetricSerializationResult serialize(
 			Map<Counter, Tuple2<QueryScopeInfo, String>> counters,
 			Map<Gauge<?>, Tuple2<QueryScopeInfo, String>> gauges,
 			Map<Histogram, Tuple2<QueryScopeInfo, String>> histograms,
-			Map<Meter, Tuple2<QueryScopeInfo, String>> meters) {
+			Map<Meter, Tuple2<QueryScopeInfo, String>> meters,
+			long maximumFramesize,
+			MetricQueryService queryService) {
 
 			buffer.clear();
+			boolean unregisterRemainingMetrics = false;
 
 			int numCounters = 0;
 			for (Map.Entry<Counter, Tuple2<QueryScopeInfo, String>> entry : counters.entrySet()) {
+				if (unregisterRemainingMetrics) {
+					queryService.unregister(entry.getKey());
+					continue;
+				}
+
 				try {
 					serializeCounter(buffer, entry.getValue().f0, entry.getValue().f1, entry.getKey());
 					numCounters++;
+					if (buffer.length() > maximumFramesize) {
+						unregisterRemainingMetrics = true;
+					}
 				} catch (Exception e) {
 					LOG.debug("Failed to serialize counter.", e);
+
 				}
 			}
 
 			int numGauges = 0;
 			for (Map.Entry<Gauge<?>, Tuple2<QueryScopeInfo, String>> entry : gauges.entrySet()) {
+				if (unregisterRemainingMetrics) {
+					queryService.unregister(entry.getKey());
+					continue;
+				}
+
 				try {
 					serializeGauge(buffer, entry.getValue().f0, entry.getValue().f1, entry.getKey());
 					numGauges++;
+					if (buffer.length() > maximumFramesize) {
+						unregisterRemainingMetrics = true;
+					}
 				} catch (Exception e) {
 					LOG.debug("Failed to serialize gauge.", e);
 				}
 			}
 
-			int numHistograms = 0;
-			for (Map.Entry<Histogram, Tuple2<QueryScopeInfo, String>> entry : histograms.entrySet()) {
-				try {
-					serializeHistogram(buffer, entry.getValue().f0, entry.getValue().f1, entry.getKey());
-					numHistograms++;
-				} catch (Exception e) {
-					LOG.debug("Failed to serialize histogram.", e);
-				}
-			}
-
 			int numMeters = 0;
 			for (Map.Entry<Meter, Tuple2<QueryScopeInfo, String>> entry : meters.entrySet()) {
+				if (unregisterRemainingMetrics) {
+					queryService.unregister(entry.getKey());
+					continue;
+				}
+
 				try {
 					serializeMeter(buffer, entry.getValue().f0, entry.getValue().f1, entry.getKey());
 					numMeters++;
+					if (buffer.length() > maximumFramesize) {
+						unregisterRemainingMetrics = true;
+					}
 				} catch (Exception e) {
 					LOG.debug("Failed to serialize meter.", e);
 				}
 			}
 
+			int numHistograms = 0;
+			for (Map.Entry<Histogram, Tuple2<QueryScopeInfo, String>> entry : histograms.entrySet()) {
+				if (unregisterRemainingMetrics) {
+					queryService.unregister(entry.getKey());
+					continue;
+				}
+
+				try {
+					serializeHistogram(buffer, entry.getValue().f0, entry.getValue().f1, entry.getKey());
+					numHistograms++;
+					if (buffer.length() > maximumFramesize) {
```

Comment:
well this naturally doesn't work since we already exceed the maximumFrameSize.

Reply:
you could throw an exception here and let the MQS decide on what to do.

Reply:
What do you mean if it exceeds the size, then throw an exception? If so, then can't we return some of the serialized metrics? Or are we just saving the exception to the MetricSerializationResult? Then judge in MQS?

Reply:
Retroactively cutting the byte-array leaks the serialization abstraction, so that's not a viable option. Something i was pondering about was to serialize each metric type into separate arrays; this would make it easy for the MQS to separate it.

Reply:
OK, I think if we adopt the strategy of throwing exception, this is the same as we judge the total size directly in MQS. Or should we think about the cost of implementation and the probability of size overflow when the metrics dump? Because I am not very clear, whether the size overflow is a high frequency scene. If not, I suggest to judge the total size directly and then return an error message. If yes, then we need to consider returning some of the metrics.

Reply:
> Retroactively cutting the byte-array leaks the serialization abstraction, so that's not a viable option. Something i was pondering about was to serialize each metric type into separate arrays; this would make it easy for the MQS to separate it.

Is this my implementation guide?

Reply:
> OK, I think if we adopt the strategy of throwing exception, this is the same as we judge the total size directly in MQS.

That is correct.

> probability of size overflow

Detailed latency metrics are the only known case where this happened which is why I'm so persistent in suggesting to drop histograms first as it eliminates this case.

> we need to consider returning some of the metrics

When you separate the byte array for each type we can always run the full serialization, and let the MQS drop histograms from the serialization result without requiring re-serialization.

Reply:
OK, I will try to separate the byte array for each type.